resolver/resolver.cc

```diff
@@ -1317,16 +1374,60 @@ class ResolveConstantsWalk {
     }
 
 public:
-    ResolveConstantsWalk() : nesting_(nullptr) {}
+    ResolveConstantsWalk() : nesting_(nullptr), loadScopeDepth_(0) {
+        localFirstDefOfManySymTable = make_unique<UnorderedMap<core::SymbolRef, core::LocOffsets>>();
+    }
 
-    void preTransformClassDef(core::Context ctx, ast::ExpressionPtr &tree) {
-        nesting_ = make_unique<Nesting>(std::move(nesting_), ast::cast_tree_nonnull<ast::ClassDef>(tree).symbol);
+    void preTransformMethodDef(core::Context ctx, ast::ExpressionPtr &tree) {
+        if (loadScopeDepth_ >= 0) {
+            loadScopeDepth_++;
+        }
+    }
+
+    void postTransformMethodDef(core::Context ctx, ast::ExpressionPtr &tree) {
+        if (loadScopeDepth_ > 0) {
+            loadScopeDepth_--;
+        }
+    }
+
+    void preTransformBlock(core::Context ctx, ast::ExpressionPtr &tree) {
+        if (loadScopeDepth_ >= 0) {
+            loadScopeDepth_++;
+        }
+    }
+
+    void postTransformBlock(core::Context ctx, ast::ExpressionPtr &tree) {
+        if (loadScopeDepth_ > 0) {
+            loadScopeDepth_--;
+        }
     }
 
     void postTransformUnresolvedConstantLit(core::Context ctx, ast::ExpressionPtr &tree) {
         walkUnresolvedConstantLit(ctx, tree);
     }
 
+    void preTransformClassDef(core::Context ctx, ast::ExpressionPtr &tree) {
+        auto &original = ast::cast_tree_nonnull<ast::ClassDef>(tree);
+        auto sym = original.symbol;
+
+        // Populate local first definitions table for out-of-order reference checking.
+        // This needs to happen *iff* the symbol is:
+        //   a). defined only in this file and nowhere else
+        //   b). defined multiple times in this file
+        //   c). this particular classDef is the first occurrence of the symbol in the file
+        if (loadScopeDepth_ == 0 && sym.isOnlyDefinedInFile(ctx.state, ctx.file)) {
```

Comment:
Do things go awry if we skip checking this in RBI files?  Do things get any faster if we skip checking this in RBI files?

Reply:
I think we should skip, I can add that in.